test/docfx.Test/TestUtility.cs

```diff
@@ -131,6 +131,44 @@ public static IDisposable EnsureFilesNotChanged(string path, bool skipInputCheck
             }
         }
 
+        public static void SendLanguageServerNotification(
+            DocfxLanguageServerTestHost testHost,
+            LanguageServerNotification notification,
+            IEnumerable<KeyValuePair<string, string>> variables = null)
+        {
+            notification.Params = ApplyVariables(notification.Params, variables);
+            testHost.SendNotification(notification);
+        }
+
+        private static JToken ApplyVariables(JToken value, IEnumerable<KeyValuePair<string, string>> variables)
+        {
+            if (variables != null && value != null)
+            {
+                if (value is JValue && value.Type == JTokenType.String)
+                {
+                    return ApplyVariables((string)value, variables);
+                }
+                else if (value is JArray array)
+                {
+                    var newArray = new JArray();
+                    foreach (var item in array)
+                    {
+                        newArray.Add(ApplyVariables(item, variables));
+                    }
+                    return newArray;
+                }
+                else if (value is JObject obj)
+                {
+                    foreach (var (key, val) in obj)
```

Comment:
create a new `JObject` similar to `JArray`, the current approach mutates input JToken.